tutorials/basic-topics/sorting/merge-sort.md

```diff
@@ -11,4 +11,519 @@ keywords:
+If right > left
+
+Find the middle point to divide the array into two halves: 
+middle = left + (right – left)/2
+Call mergeSort again for first half for further dividing:   
+Call mergeSort(array, left, middle)
+Call mergeSort again  for second half for further dividing:
+Call mergeSort(array, middle+ 1, right)
+Merge the two halves sorted in steps 2 and 3:
+Call merge(arrary, left, middle, right)
+```
+
+<Tabs>
+<TabItem value="cpp" label="C++">
+
+```cpp
+// call mergeSort(nums, 0, nums.size() - 1) to sort 
+
+void merge(vector<int>& nums, int l, int m, int r){
+vector<int> tmp(r - l + 1); // create a temporary array
+int i = l; // index for left subarray
+int j = m + 1; // index for right subarray
+int k = 0; // index for temporary array
+while(i <= m && j <= r){
+    if(nums[i] <= nums[j]) tmp[k++] = nums[i++];   
+    
+    // increment the left pointer if the right pointer element is bigger since we are sorting in ascending order ,left(smaller element) goes first
+    
+    else tmp[k++] = nums[j++];
+}
+while(i <= m) tmp[k++] = nums[i++];
+while(j <= r) tmp[k++] = nums[j++]; 
+for(i = 0; i < k; i++) nums[l + i] = tmp[i];
+}
+
+
+void mergeSort(vector<int>& nums, int l, int r){
+if(l >= r) return;
+int m = l + (r - l) / 2; //middle index, same as (l+r)/2
+mergeSort(nums, l, m);
+mergeSort(nums, m + 1, r);
+merge(nums, l, m, r);
+}
+
+```
+
+</TabItem>
+
+<TabItem value="py" label="Python">
+
+```py
+def merge_Sort(nums): 
+if len(nums) > 1: 
+    mid = len(nums)//2   # calculate the middle to divide by mid
+    leftArray = nums[:mid]    #slicing 
+    rightArray = nums[mid:]  
+
+    merge_Sort(leftArray)
+    merge_Sort(rightArray)
+
+    i = j = k = 0
+
+    while i < len(leftArray) and j < len(rightArray): 
+        if leftArray[i] < rightArray[j]: 
+            nums[k] = leftArray[i] 
+            i+=1 //incrementing the smaller element side pointer
+        else: 
+            nums[k] = rightArray[j] 
+            j+=1
+        k+=1
+
+    while i < len(leftArray): 
+        nums[k] = leftArray[i] 
+        i+=1
+        k+=1
+
+    while j < len(rightArray): 
+        nums[k] = rightArray[j] 
+        j+=1
+        k+=1
+
+```
+
+</TabItem>
+<TabItem value="java" label="Java">
+
+```java
+// call mergeSort(nums , 0 ,nums.length -1) in public
+
+private void mergeSort(int[] nums, int l, int r) {   // nums -> ArrayList , l -> starting index , r-> end of the ArrayList
+    if (l >= r) return;
+    int mid = l + (r - l) / 2;
+    mergeSort(nums, l, mid);
+    mergeSort(nums, mid + 1, r);
+    merge(nums, l, r);
+}
+private void merge(int[] nums, int l, int r) {
+    int mid = l + (r - l) / 2;
+    int[] tmp = new int[r - l + 1];
+    int i = l, j = mid + 1, k = 0;
+    while (i <= mid || j <= r) {
+        if (i > mid || j <= r && nums[i] > nums[j]) {
+            tmp[k++] = nums[j++];
+        } else {
+            tmp[k++] = nums[i++];
+        }
+    }
+```
+</TabItem>
+</Tabs>
+
+
+## Bottom Up Approach /Iterative technique
```

Comment:
Approach / Iterative

Reply:
Still not ok, after / give a space like this.. Approach / Iterative

![image](https://user-images.githubusercontent.com/10242660/193452927-18f7dcb4-cb19-410f-9e37-9a42ca7800fa.png)